File: packages/javascript/src/internal/drawing-state.js
import { DrawingError, normalizeCommand, validateEnvironment } from "./drawing.js";

const MAX_BATCH_COMMANDS = 4_096;
const MAX_FRAME_COMMANDS = Number.MAX_SAFE_INTEGER;
const BEGIN_FAILURE_CODES = new Set(["UNSUPPORTED_CAPABILITY", "RESOURCE_FAILURE", "RENDER_FAILURE"]);

/** A fresh-raster lifecycle error with an optional absolute command index. */
export class FrameError extends DrawingError {
  constructor(code, commandIndex = null) {
    super(code);
    this.name = "FrameError";
    this.commandIndex = commandIndex;
  }
}

function frameError(error, commandIndex = null) {
  if (error instanceof FrameError) return error;
  if (error instanceof DrawingError) return new FrameError(error.code, commandIndex);
  return new FrameError("INVALID_STATE");
}

function freezeRecord(value) {
  if (Array.isArray(value)) {
    for (const item of value) freezeRecord(item);
    return Object.freeze(value);
  }
  if (value !== null && typeof value === "object") {
    for (const item of Object.values(value)) freezeRecord(item);
    return Object.freeze(value);
  }
  return value;
}

function immutablePlan(value) {
  return freezeRecord(value);
}

function validBatchFailureOffset(offset, inputCount) {
  return offset === null || (typeof offset === "number" && Number.isSafeInteger(offset) &&
    offset >= 0 && offset < inputCount);
}

/**
 * Pure internal lifecycle state for a fresh-raster frame.
 * It owns no renderer, surface lease, callback, or producer command buffer.
 */
export class FrameState {
  #state = "new";
  #count = 0;
  #environment = null;
  #pending = null;
  #pendingKind = null;

  get state() {
    return this.#state;
  }

  get count() {
    return this.#count;
  }

  #abort() {
    this.#environment = null;
    this.#pending = null;
    this.#pendingKind = null;
    this.#state = "aborted";
  }

  #invalidState() {
    if (this.#state !== "completed") this.#abort();
    throw new FrameError("INVALID_STATE");
  }

  #requirePending(plan, kind) {
    const expectedState = kind === "begin" ? "new" : "active";
    if (this.#state !== expectedState || this.#pending !== plan || this.#pendingKind !== kind) {
      this.#invalidState();
    }
  }

  /** Validate an environment and issue an immutable begin plan. */
  prepareBegin(environment) {
    if (this.#state !== "new" || this.#pending !== null) this.#invalidState();
    const reservation = immutablePlan({});
    this.#pending = reservation;
    this.#pendingKind = "begin-preparing";
    try {
      const detachedEnvironment = immutablePlan(validateEnvironment(environment));
      const plan = immutablePlan({ environment: detachedEnvironment });
      if (this.#state !== "new" || this.#pending !== reservation) throw new FrameError("INVALID_STATE");
      this.#pending = plan;
      this.#pendingKind = "begin";
      return plan;
    } catch (error) {
      this.#abort();
      throw frameError(error);
    }
  }

  /** Resolve a matching begin plan after adapter setup succeeds. */
  activate(plan) {
    this.#requirePending(plan, "begin");
    this.#environment = plan.environment;
    this.#pending = null;
    this.#pendingKind = null;
    this.#state = "active";
  }

  /** Resolve a matching begin plan after a phase-specific adapter failure. */
  failBegin(plan, code) {
    this.#requirePending(plan, "begin");
    if (!BEGIN_FAILURE_CODES.has(code)) this.#invalidState();
    this.#abort();
    throw new FrameError(code);
  }

  /**
   * Fully validate and normalize one bounded batch before any adapter drawing.
   * Slots retain input offsets, including no-op commands, until synchronous resolution.
   */
  prepareBatch(commands) {
    if (this.#state !== "active" || this.#pending !== null) this.#invalidState();
    const reservation = immutablePlan({});
    this.#pending = reservation;
    this.#pendingKind = "batch-preparing";
    try {
      if (!Array.isArray(commands)) throw new FrameError("INVALID_BATCH");
      const inputCount = commands.length;
      if (inputCount > MAX_BATCH_COMMANDS || this.#count > MAX_FRAME_COMMANDS - inputCount) {
        throw new FrameError("INVALID_BATCH");
      }
      const snapshot = new Array(inputCount);
      for (let offset = 0; offset < inputCount; offset += 1) {
        snapshot[offset] = commands[offset];
      }
      if (this.#state !== "active" || this.#pending !== reservation) throw new FrameError("INVALID_STATE");

      const baseIndex = this.#count;
      const slots = new Array(inputCount);
      for (let offset = 0; offset < inputCount; offset += 1) {
        try {
          slots[offset] = immutablePlan({ sourceOffset: offset, command: normalizeCommand(snapshot[offset], this.#environment) });
        } catch (error) {
          throw frameError(error, baseIndex + offset);
        }
      }
      if (this.#state !== "active" || this.#pending !== reservation) throw new FrameError("INVALID_STATE");
      const plan = immutablePlan({ baseIndex, inputCount, slots });
      if (this.#state !== "active" || this.#pending !== reservation) throw new FrameError("INVALID_STATE");
      this.#pending = plan;
      this.#pendingKind = "batch";
      return plan;
    } catch (error) {
      this.#abort();
      throw frameError(error);
    }
  }

  /** Commit a matching batch only after the adapter completes every slot. */
  commitBatch(plan) {
    this.#requirePending(plan, "batch");
    this.#count += plan.inputCount;
    this.#pending = null;
    this.#pendingKind = null;
  }

  /** Abort a matching batch after a native failure at an original input offset. */
  failBatch(plan, sourceOffset) {
    this.#requirePending(plan, "batch");
    if (!validBatchFailureOffset(sourceOffset, plan.inputCount) ||
        sourceOffset !== null && plan.slots[sourceOffset].command.outcome === "noop") {
      this.#invalidState();
    }
    const commandIndex = sourceOffset === null ? null : plan.baseIndex + sourceOffset;
    this.#abort();
    throw new FrameError("RENDER_FAILURE", commandIndex);
  }

  /** Issue an identity-bound end token for adapter finalization. */
  prepareEnd() {
    if (this.#state !== "active" || this.#pending !== null) this.#invalidState();
    const reservation = immutablePlan({});
    this.#pending = reservation;
    this.#pendingKind = "end-preparing";
    try {
      const token = immutablePlan({});
      if (this.#state !== "active" || this.#pending !== reservation) throw new FrameError("INVALID_STATE");
      this.#pending = token;
      this.#pendingKind = "end";
      return token;
    } catch (error) {
      this.#abort();
      throw frameError(error);
    }
  }

  /** Resolve a matching end token and mark the frame completed. */
  completeEnd(token) {
    this.#requirePending(token, "end");
    this.#pending = null;
    this.#pendingKind = null;
    this.#environment = null;
    this.#state = "completed";
  }

  /** Resolve a matching end token after native finalization fails. */
  failEnd(token) {
    this.#requirePending(token, "end");
    this.#abort();
    throw new FrameError("RENDER_FAILURE");
  }

  /** Idempotently abort an unfinished frame; completed frames remain immutable. */
  abort() {
    if (this.#state === "completed") throw new FrameError("INVALID_STATE");
    this.#abort();
  }

  /** Test-only exact-count hook; this internal module is never package-exported. */
  static forTestActive(environment, count) {
    if (typeof count !== "number" || !Number.isSafeInteger(count) || count < 0 || count > MAX_FRAME_COMMANDS) {
      throw new FrameError("INVALID_STATE");
    }
    const state = new FrameState();
    const plan = state.prepareBegin(environment);
    state.activate(plan);
    state.#count = count;
    return state;
  }
}

File: packages/python/procedurals/_drawing_state.py
"""Internal portable frame lifecycle; native surface ownership stays in adapters.

drawing.fresh-raster-2d v0.1.0, motivated by pelines independent marks and
ciserp ordered path marks. No renderer, asset, callback or surface is retained.
"""
from dataclasses import dataclass
from types import MappingProxyType
from ._drawing import DrawingError, normalize_command, validate_environment


class FrameError(DrawingError):
    def __init__(self,code,command_index=None):
        self.command_index=command_index
        super().__init__(code)


def _immutable(value):
    if isinstance(value,dict):return MappingProxyType({k:_immutable(v) for k,v in value.items()})
    if isinstance(value,list):return tuple(_immutable(v) for v in value)
    return value


@dataclass(frozen=True,eq=False,slots=True)
class _Plan:
    kind:str
    environment:object=None
    base_index:int=0
    input_count:int=0
    slots:tuple=()


class FrameState:
    """Synchronous two-phase lifecycle with private identity-based pending plans."""
    def __init__(self):
        self._state='new'
        self._count=0
        self._environment=None
        self._pending=None

    @property
    def state(self):return self._state

    @property
    def count(self):return self._count

    def _error(self,code,index=None):
        if self._state!='completed':
            self._state='aborted'
            self._pending=None
            self._environment=None
        raise FrameError(code,index)

    def _reserve(self,state):
        if self._state!=state or self._pending is not None:self._error('INVALID_STATE')
        token=object();self._pending=token
        return token

    def _publish(self,reservation,plan,state):
        if self._state!=state or self._pending is not reservation:self._error('INVALID_STATE')
        self._pending=plan
        return plan

    def _match(self,plan,kind,state):
        if self._state!=state or self._pending is not plan or not isinstance(plan,_Plan) or plan.kind!=kind:
            self._error('INVALID_STATE')

    def prepare_begin(self,environment):
        reserved=self._reserve('new')
        try:env=_immutable(validate_environment(environment))
        except FrameError:self._error('INVALID_STATE')
        except DrawingError:self._error('INVALID_ENVIRONMENT')
        except BaseException:
            self.abort();raise
        return self._publish(reserved,_Plan('begin',environment=env),'new')

    def activate(self,plan):
        self._match(plan,'begin','new')
        self._environment=plan.environment
        self._pending=None;self._count=0;self._state='active'

    def fail_begin(self,plan,code):
        self._match(plan,'begin','new')
        if code not in ('UNSUPPORTED_CAPABILITY','RESOURCE_FAILURE','RENDER_FAILURE'):self._error('INVALID_STATE')
        self._error(code)

    def prepare_batch(self,commands):
        reserved=self._reserve('active')
        try:
            # Passive built-in containers only; snapshot within the bounded budget
            # before any command accessor can change the caller's original list.
            if type(commands) not in (list,tuple) or len(commands)>4096 or self._count>9007199254740991-len(commands):
                self._error('INVALID_BATCH')
            commands=tuple(commands)
            slots=[]
            for offset,command in enumerate(commands):
                try:slots.append(_immutable(normalize_command(command,self._environment)))
                except FrameError:self._error('INVALID_STATE')
                except DrawingError:self._error('INVALID_COMMAND',self._count+offset)
            return self._publish(reserved,_Plan('batch',base_index=self._count,input_count=len(commands),slots=tuple(slots)),'active')
        except FrameError:raise
        except BaseException:
            self.abort();raise

    def commit_batch(self,plan):
        self._match(plan,'batch','active')
        self._count=plan.base_index+plan.input_count
        self._pending=None

    def fail_batch(self,plan,source_offset=None):
        self._match(plan,'batch','active')
        if source_offset is not None and (type(source_offset) is not int or not 0<=source_offset<plan.input_count):
            self._error('INVALID_STATE')
        if source_offset is not None and plan.slots[source_offset]['outcome']=='noop':self._error('INVALID_STATE')
        self._error('RENDER_FAILURE',None if source_offset is None else plan.base_index+source_offset)

    def prepare_end(self):
        reservation=self._reserve('active')
        return self._publish(reservation,_Plan('end'),'active')

    def complete_end(self,plan):
        self._match(plan,'end','active')
        self._pending=None;self._environment=None;self._state='completed'

    def fail_end(self,plan):
        self._match(plan,'end','active');self._error('RENDER_FAILURE')

    def abort(self):
        if self._state=='completed':self._error('INVALID_STATE')
        self._pending=None;self._environment=None;self._state='aborted'
